cmd/shell_test.go

```diff
@@ -228,3 +229,57 @@ func TestFailedConfigValidation(t *testing.T) {
 	test.AssertError(t, err, "Expected validation error")
 	test.AssertContains(t, err.Error(), "'required'")
 }
+
+func TestFailExit(t *testing.T) {
+	// Test that when Fail is called with a `defer AuditPanic()`,
+	// the program exits with a non-zero exit code and logs
+	// the result (but not stack trace).
+	// Inspired by https://go.dev/talks/2014/testing.slide#23
+	if os.Getenv("TIME_TO_DIE") == "1" {
+		defer AuditPanic()
+		Fail("tears in the rain")
+		return
+	}
+
+	// gosec points out that os.Args[0] is tainted, but we only run this as a test
+	// so we are not worried about it containing an untrusted value.
+	//nolint:gosec
+	cmd := exec.Command(os.Args[0], "-test.run=TestFailExit")
+	cmd.Env = append(os.Environ(), "TIME_TO_DIE=1")
+	output, err := cmd.CombinedOutput()
+	test.AssertError(t, err, "running a failing program")
+	test.AssertContains(t, string(output), "[AUDIT] tears in the rain")
+	// "goroutine" usually shows up in stack traces, so we check it
+	// to make sure we didn't print a stack trace.
+	test.AssertNotContains(t, string(output), "goroutine")
+}
+
+func testPanicStackTraceHelper() {
+	var x *int
+	*x = 1
+}
+
+func TestPanicStackTrace(t *testing.T) {
+	// Test that when a nil pointer dereference is hit after a
+	// `defer AuditPanic()`, the program exits with a non-zero
+	// exit code and prints the result (but not stack trace).
+	// Inspired by https://go.dev/talks/2014/testing.slide#23
```

Comment:
This is gnarly, so the test is going to call another copy of itself with the environment variable that will trigger the failure scenario.